### DISCUSSION TITLE
Hard to understand "Migrating To on_attach"

### DISCUSSION THREAD

**Discussion by smprather** [nvim-tree/nvim-tree.lua#2224] [Q&A]
I'm getting this message on nvim startup:

```
view.mappings.list has been deprecated in favour of on_attach. Please run :NvimTreeGenerateOnAttach
and visit https://github.com/nvim-tree/nvim-tree.lua/wiki/Migrating-To-on_attach
```

So I'm trying to read the migration instructions. Forgive me for being dense, but I'm finding it difficult to understand. The code snippets don't mention which files they are in. I don't understand "Add that function to your setup lua". What function? Oh, the "on_attach" function in step 0? Step 0 was never explained, except "generate an on_attach function". I don't mind having a passing copy/paste knowledge of Lua, but I'd rather not learn it beyond that (I have no other use for it other than nvim).

As I type this question, I'm kind of figuring out the answer. But can there perhaps be a better user experience? I don't mind starting over from scratch. I don't have any custom mappings. Sorry to be a whiner.

**Comment by smprather** [comment]
I figured it out and got it working. So my question is reduced to "perhaps be a better user experience". Or maybe the (perceived) complexity is required?

**Reply by gegoune** [reply]
Have you got any suggestions how documentation/guide could be improved?

**Reply by alex-courtis** [reply]
The migration is a complex page.

Rather than focusing on the past and the migration page, we can focus on the future.

`:help nvim-tree-mappings` is a lot simpler, with a single block of code to paste. @smprather could you please review that as well?

**Reply by alex-courtis** [reply]
Would it be clearer if we put `:help nvim-tree-mappings` at the start of the migration page?

**Reply by smprather** [reply]
Sorry to complain and then not participate in the solution.

Unfortunately, I have spent a couple of days diving DEEP into the nvim ecosystem (and, therefor, Lua). So what was once confusing is not so much anymore.

I just read `:help nvim-tree-mappings`. I would say the only thing slightly confusing is that the on_attach hook is normally associated with an LSP (true?false?).

Anyway, I think we all have bigger fish to deal with. Thanks @alex-courtis !

**Comment by back2childhood** [comment]
I still don't know how to solve this problem ...... /(ㄒoㄒ)/  So I use nvim-tree with this annoying hint now.

**Reply by alex-courtis** [reply]
Perhaps we could put a small, optional `on_attach` in quickstart.

**Reply by alex-courtis** [reply]
We could do this. It's very long now, not exactly a quick start...

```lua
  -- disable netrw at the very start of your init.lua (strongly advised)
  vim.g.loaded_netrw = 1
  vim.g.loaded_netrwPlugin = 1

  -- set termguicolors to enable highlight groups
  vim.opt.termguicolors = true

  -- optional on_attach to modify mappings
  local function my_on_attach(bufnr)
    local api = require('nvim-tree.api')

    local function opts(desc)
      return { desc = 'nvim-tree: ' .. desc, buffer = bufnr, noremap = true, silent = true, nowait = true }
    end

    -- default mappings
    api.config.mappings.default_on_attach(bufnr)

    -- custom mappings
    vim.keymap.set('n', '<C-t>', api.tree.change_root_to_parent,        opts('Up'))
    vim.keymap.set('n', '?',     api.tree.toggle_help,                  opts('Help'))
  end

  -- empty setup using defaults
  require("nvim-tree").setup()

  -- OR setup with some options
  require("nvim-tree").setup({
    sort_by = "case_sensitive",
    view = {
      width = 30,
    },
    renderer = {
      group_empty = true,
    },
    filters = {
      dotfiles = true,
    },
    on_attach = my_on_attach, -- optional
  })
```
